### DISCUSSION TITLE
Allow # tags in note body

### DISCUSSION THREAD

**Discussion by garyo** [ahyatt/ekg#105] [Ideas]
Coming from logseq, I expect to be able to just type #tags anywhere in the #ekg note body. So I wrote a before-save hook to extract all such tags from the body and add them to the tags list. Here it is, in case you want something similar.
(Extra credit if you can figure out why the `(setf ...)` form doesn't work in here!)

```emacs-lisp
  (defun get-ekg-body-tags (note)
    "Get #tags from body of EKG note"
    (let* ((string (ekg-note-text note))
           (regexp "#\\([-_.a-zA-Z0-9]+\\)")
           matches
           (newtags (save-match-data
                      (let ((pos 0)
                            matches)
                        (while (string-match regexp string pos)
                          (push (match-string 1 string) matches)
                          (setq pos (match-end 0)))
                        matches))))
      (seq-uniq (append newtags (ekg-note-tags note)))))

  (defun my-ekg-note-pre-save-hook (note)
    "Apply #tags found in body to the note's tags"
    (let ((tags (get-ekg-body-tags note)))
      (message "Setting tags to %s" tags)
      ;; Workaround: the setf macro below doesn't work properly;
      ;; it macroexpands to a call to a function named
      ;; "(setf ekg-note-tags)" including the parens and spaces!
      ;; Just call aset to set the slot instead.
      ;; See https://emacs.stackexchange.com/questions/79007
      ;; (setf (ekg-note-tags note) tags)
      (aset note (cl-struct-slot-offset 'ekg-note 'tags) tags)
      (ekg--normalize-note note)
    ))

  ;; Allow a note to have tags in the body, by scanning the body before saving and adding any tags to the note's tags.
  (add-hook 'ekg-note-pre-save-hook 'my-ekg-note-pre-save-hook)
```

**Comment by ahyatt** [comment]
That's great!  I've also thought about having tag completion, which would help, and at the end of the completion you could add it to the tags.  You might need both, though, because if you don't use completion we wouldn't detect you added it.

One thing that worries me is that you can change remove a tag in the text when editing an existing note, and tags you previously had because of that text wouldn't be removed.  Maybe it's not such a big deal, though.

**Reply by garyo** [reply]
Personally I'm OK with it not auto-removing tags if you remove them from the text, but that would be a nice feature to add. One way would be for tags to have properties; tags marked "explicit" wouldn't be synced with tags from the text, and other tags would be. But that's a more significant change.

**Comment by ahyatt** [comment]
I've taken a different approach to this in the `inline-tags` branch.  The idea is that you can just complete tags by typing "#", and completing the tag, and it will just add the tag when completed to the note properties (and run the various hooks involved).  Please check it out and provide any feedback!

**Comment by garyo** [comment]
I just tried this branch. It works well for existing tags, but it didn't seem to add #newtags (new inline tags) when I added them #inline like that, even if I triggered completion after typing them (naturally it says no completion found), unless I enable my save-hook above. I'm using `company-complete` if that matters.
This is more visible when trying to add new hierarchical tags; if I already have `#proglang/cpp` and I want to add a new tag `#proglang/elisp`, if I type `#proglang/` and `M-RET`, it eagerly auto-completes to the existing tag and adds it. 
I'd prefer a UX that doesn't require me to know if a tag has been already used or not.

**Comment by ahyatt** [comment]
Yes, this is a good point. I'm not able to get new tags to work either.  Perhaps a combination of both would be good - I like the idea of being able to complete tags.  One question, though, since tags can have spaces in ekg, we wouldn't know by default when a tag has finished.  Your solution works for single word tags, and maybe that's the best we can do if we have to detect tags, or maybe we can select something like "#[multiple world tag]".

Another variation I'd like to think about: maybe instead of using completion we can have a command that will look for text between "#" and the point, and insert that as a tag in the note.  Let me play around with these approaches and see what might work well.

**Reply by garyo** [reply]
Completion is definitely useful! I'm glad you're adding that, no matter what else gets done. But if there is some better way to scan the body for tags, then completion could just do what it normally does -- complete the tag name.

As for multi-word tags, logseq uses `[[multi word tag]]` but I kind of like your `#[multi word tag]`; either would be OK with me.

In any case I will probably continue to use my hook because there are lots of ways for text to get into a note (e.g. copy/paste) and I would never want body tags to get missed. Ekg could add a function/command for this, as you suggest (`ekg-apply-body-tags` or something), which could be run as a save hook or periodically when the buffer changes.

**Reply by ahyatt** [reply]
I added another commit just now to the branch that changes it to be similar to what you have done.  When you do inline completion, it will complete the tag, and if there is a space, it will enclose the tag in brackets.  But it doesn't add the tag to the tags yet, it just completes the text.  When saved, the tags in the text will be added to the note tags, which is similar to your method, but with the addition of accepting the bracketed tags as well.

**Reply by ahyatt** [reply]
I've now merged my branch in with the develop branch, and made further improvements, notably making inline tags links in org-mode.  This will be in the next major release.
